### PR TITLE
Add ISmartContractBlockHeader support to PosBlockHeader

### PR DESCRIPTION
**Pull request by quantumagi**
This is a backwards compatible change that will not change the way `PosBlockHeader` is serialized unless the appropriate version bit is set to indicate that additional fields required by system contracts should be serialized. Serialization of `ProvenBlockHeader` , which derives from `PosBlockHeader`, is similarly affected by this change.

This PR is being split off from PR #450 to reduce the size of that PR.

### CODE REVIEW

File: src/NBitcoin/BlockStake.cs [hunk 1]
```diff
@@ -270,16 +270,57 @@ public override Transaction CreateTransaction(byte[] bytes)
...
         public override int CurrentVersion => 7;
 
+        private ushort extendedHeaderSize => (ushort)(hashStateRootSize + receiptRootSize + this.logsBloom.GetCompressedSize());
+
+        /// <summary>
+        /// Root of the state trie after execution of this block. 
+        /// </summary>
+        private uint256 hashStateRoot;
+        public uint256 HashStateRoot { get { return this.hashStateRoot; } set { this.hashStateRoot = value; } }
+        private static int hashStateRootSize = (new uint256()).GetSerializeSize();
```
Comment (fassadlr) [inline src/NBitcoin/BlockStake.cs:302]: Question, does this represent a "proper" `uint256`s size? i.e, not one that is "empty"... ?

File: src/NBitcoin/BlockStake.cs [hunk 2]
```diff
@@ -323,6 +364,48 @@ public override uint256 GetPoWHash()
                 return HashX13.Instance.Hash(ms.ToArray());
             }
         }
+
```
Comment (fassadlr) [inline src/NBitcoin/BlockStake.cs:367]: Extra line ;)

File: src/NBitcoin/BlockStake.cs [hunk 3]
```diff
@@ -270,16 +270,57 @@ public override Transaction CreateTransaction(byte[] bytes)
...
     /// <summary>
     /// A POS block header, this will create a work hash based on the X13 hash algos.
     /// </summary>
 #pragma warning disable 618
-    public class PosBlockHeader : BlockHeader
+    public class PosBlockHeader : BlockHeader, ISmartContractBlockHeader
 #pragma warning restore 618
     {
+        // Determines whether this object should serialize the new fields associated with smart contracts.
+        public bool HasSmartContractFields => (this.version & 0x10000000 /* ExtendedHeaderBit */) != 0;
```
Comment (fassadlr) [inline src/NBitcoin/BlockStake.cs:290]: Can we not reference the extended header bit from the ThresholdConditionCache? Or vice versa?

**Review by fassadlr — COMMENTED** [review]
Looks good, just a few questions.

File: src/NBitcoin/BlockStake.cs (hunk 1 above)
Comment (quantumagi) [reply src/NBitcoin/BlockStake.cs:302]: `uint256` is always 32 bytes.

File: src/NBitcoin/BlockStake.cs (hunk 3 above)
Comment (quantumagi) [reply src/NBitcoin/BlockStake.cs]: Not from here but I've moved the definition.

Comment (quantumagi) [reply src/NBitcoin/BlockStake.cs]: Not from here but I've moved the definition.

**Review by fassadlr — APPROVED** [review]

**Comment by fassadlr** [comment]
Once CI passes you can merge :)

**Pull request merged**